### ISSUE TITLE
Incorrect handling of null character in strings

### ISSUE THREAD

**Issue by angleKH** [react-native-async-storage/async-storage#599]
## Current behavior

Either `AsyncStorage.setItem` or  `AsyncStorage.getItem` is failing to handle strings with null characters. When reading a string value stored with `AsyncStorage.setItem` that contains a null character, only the string up to the first null character is read.

## Expected behavior

The string that was read should be the same as the string that was stored, and the string should be stored properly.

## Repro steps
Example, first, store with:
```js
const stringToStore = ["a","b","c","d"].join("\0"); // equivalent to "a\u0000b\u0000c\u0000d"
AsyncStorage.setItem("@someKey", stringToStore).catch(console.error);
```
Then read with:
```js
AsyncStorage.getItem("@someKey").then(s => { console.log("Read "+JSON.stringify(s)); }).catch(console.error);
```
The console will print out `Read "a"` instead of `Read "a\u0000b\u0000c\u0000d"` or similar.
## Environment

- **Platforms tested**:
  - Android
- **AsyncStorage version**: `^1.15.4`
- **Expo version**: `~41.0.0`

**Comment by tido64** [comment]
This sounds like a real issue. @Krizzu, any insights?

**Comment by Krizzu** [comment]
Well, it's kinda tricky to point a blame part here - if you console log `"a\u0000b\u0000c\u0000d"`, you'll get `abcd`, I suspect due to log trimming null characters.

You can check the length of a value before saving and after reading from AsyncStorage - maybe it's terminated at the bridge, due to `\0` being used to terminate string in C?

**Comment by angleKH** [comment]
I used `console.log` on the result of `JSON.stringify` with the string so that non-printable characters with no visual representation within the string such as the null character can be seen as an escape sequence for demonstration purposes. Strings in JS are not null-terminated and are required by the specification to support the null character. The string is definitely truncated somewhere in between the calls to `AsyncStorage.setItem` and `AsyncStorage.getItem`.

After looking around a bit, I heard this project uses SQLite for Android. If this is the case, then the API for AsyncStorage would have inconsistent behavior between Android and other platforms AsyncStorage don't use SQLite in its implementation for such as iOS. I found this page, which perhaps may be useful: https://sqlite.org/nulinstr.html

I'm not sure where the cause is. It's possible this could be an upstream issue if this project uses another library for interacting with SQLite. The only C++ sources in this repository seem to be for the Windows platform, so C strings being null-terminated seems unlikely as the cause; at least for this project.
